proposals/per-extension-language-preferences.md

```diff
@@ -0,0 +1,251 @@
+- [Optional] Extensions could integrate third-party i18n implementations (frameworks) with the browser's built-in language selection menu.
+
+### Known Consumers
+
+This is a generic i18n feature, which means that all types of extensions can use it. Some extensions already provide this functionality via workarounds or non-browser.i18n implementations.
+
+## Specification
+
+### Definitions
+
+In this document:
+- **locale** or **language code** or **language tag** is a string that represent a language, defined in [BCP 47](https://www.rfc-editor.org/info/bcp47).
+For example, `en-US`, `zh-CN`, `fr`. It is used by `browser.i18n`, `Date`, `Intl` and various other APIs.
+- **the user preferred language** is the language set by `i18n.setCurrentLanguage(code)`. The user may not have set it.
+- **the extension displayed language** is the language which the extension is displayed in. Note that the extension displayed language may be different from the browser UI language. For example, if the user preferred language is not set, the browser UI language is English, and the extension only supports French (default) and Japanese, then the extension displayed language is French because `i18n.getMessage()` returns French message at this situation.
+
+### Schema
+
+```ts
+/**
+ * Get the extension displayed language.
+ * return the language tag, e.g. en-US.
+ */
+i18n.getCurrentLanguage() : string
+
+/**
+ * Set the user preferred language.
+ * if code is null, revert to the unset state.
+ * if code is not valid, return a rejected Promise.
+ * else return a Promise resolved when the operation is complete.
+ */
+i18n.setCurrentLanguage(code: string) : Promise<void>
+
+/**
+ * Get all languages that supported by this extension.
+ * return a Promise resolved with an array of language tags.
+ */
+i18n.getAllLanguages(): Promise<string[]>
+
+/**
+ * After changing to a new language, the browser triggers a language changed event.
+ * The callback is function(code: string) : void.
+ * The code parameter in the callback is the new language.
+ */
+i18n.onLanguageChanged.addListener(callback)
+```
+
+### Behavior
+
+#### Content Scripts Availability
+
+In history, all `browser.i18n` APIs were available in content scripts. But some of methods in this proposal should not be available in content scripts because of their purpose.
+
+| Method  | Content Scripts Availability |
+| ------------- | ------------- |
+| `getCurrentLanguage()` | Yes |
+| `setCurrentLanguage()` | No |
+| `getAllLanguages()` | No |
+| `onLanguageChanged()` | Yes |
+
+#### A New Predefined Message `@@current_locale`
+There is an exsiting predefined message `@@ui_locale`, that reflects the value of `i18n.getUILanguage()`, but the value uses underscore (e.g. "en_US") as separator in Chrome and hyphen (e.g. "en-US") in Firefox.
+
+Relative to `@@ui_locale`, a new predefined message, ``@@current_locale``, should be added, which reflects the value of `i18n.getCurrentLanguage()`.
```

Comment:
This is a late addition to the proposal. What use cases are served with the proposed `@current_locale`? This property does not automatically follow from the three listed use cases in the proposal document.

Additionally, because the stylesheet does not update automatically in response to UI language changes, the use of the proposed `@current_locale` in the stylesheet can result in UI getting out of sync with the actual language, unless the extension has additional code to deal with language changes (with `i18n.onLanguageChanged`). If an extension already needs to account for this, then they may as well do all language-dependent CSS from JavaScript. These days, they can use CSS variables to control stylesheet behavior in a more powerful way than `@`-macros can. For example, the only way for an extension to reload the stylesheet to read the new `@` values is to replace the style sheet element with a new one that has the same URL and a random extra addition (query string) to bust the style cache.

Because of this, I am inclined to vote for removing the proposed `@current_locale` from the proposal.

Reply:
Personally, I don't pursue this predefined message very much, but only add it for completeness reason. @carlosjeurissen Would you mind deleting this part? Because you suggested it.

Reply:
Also not feeling too strongly about it. One of the motivations is described here:
https://github.com/w3c/webextensions/issues/642#issuecomment-2181397837

As a way to get the path of the language files in JavaScript for example.

In general I agree with not using the predefined messages in CSS. Personally not been using them in any shape or form in the past few years in any projects in CSS. It very much feels like a hack and we might want to deprecate some of them moving forward.